Improve attribute range handling for attributed function types in sel…

quanzhuo opened the pull request:
…ection

This fix https://github.com/clangd/clangd/issues/2488

quanzhuo committed:
Improve attribute range handling for attributed function types in selection

This fix https://github.com/clangd/clangd/issues/2488

HighCommander4 reviewed (approved):
Thanks for the patch! Looks great, I just have a suggestion for one additonal test case.

HighCommander4 reviewed (changes requested):
CI is showing the test is failing, that will need to be fixed.

quanzhuo committed:
Add test for AttributedTypeLoc

quanzhuo commented:
Hi @HighCommander4 

Thanks for the review! I've added the test for selecting `AttributedTypeLoc`, when the cursor is on the attribute. However, I'm having trouble with the second test case I added. It's failing with a range mismatch:

```
D:\repos\llvm-project\clang-tools-extra\clangd\unittests\SelectionTests.cpp:635
Expected equality of these values:
  nodeRange(T.commonAncestor(), AST)
    Which is: 0:17-0:57
  Test.range()
    Which is: 0:17-0:60
struct X { const [[int* Foo() const <:[clang::life^timebound]:>]] {return nullptr;}; };
 TranslationUnitDecl
   CXXRecordDecl struct X {}
     CXXMethodDecl const int *Foo() const
      .AttributedTypeLoc const int *() const [[clang::lifetimebound]]
```

The issue appears to be that the test annotation parser stops at the `]` in `]:>`, which truncates the expected range marker.

Thanks for your guidance!

HighCommander4 commented:
> The issue appears to be that the test annotation parser stops at the `]` in `]:>`, which truncates the expected range marker.

It looks like the problem is not with the expected range marker (this is `Test.range()`, which is `0:17-0:60`, which looks correct), but with the actual range of the `AttributedTypeLoc` node, which is `0:17-0:57`.

It looks like the AST is reporting that the `AttributedTypeLoc` node ends at the `lifetimebound` token, rather than the `:>` token. This is a bug in the AST modeling of `AttributedTypeLoc` (or perhaps, based on [this comment](https://searchfox.org/llvm/rev/c8187f6539166a1d520c7f6c4f6e8857ae56a905/clang/lib/AST/TypeLoc.cpp#581-589), a deliberate quirk).

We can revise the test to expect the node to end there, i.e.:

```
struct X { const [[int* Foo() const <:[clang::life^timebound]]]:>
```

though I find it more readable to use the digraph for both brackets:

```
struct X { const [[int* Foo() const <:<:clang::life^timebound]]:>:>
```

and maybe add a comment saying that this is a quirk in the AST modeling of `AttributedTypeLoc`.

quanzhuo committed:
Update test for AttributedTypeLoc

quanzhuo commented:
Hi @HighCommander4,

Thank you for the detailed explanation!  I've updated the test case and comment as you suggested. The test should pass now. Thanks again for your patient guidance!

HighCommander4 reviewed (approved):
Thanks, LGTM!
